misc/formatter/modules/formatter-core/src/main/java/org/ballerinalang/formatter/core/FormatterUtils.java

```diff
@@ -244,15 +290,90 @@ static Token formatToken(Token token, int leadingSpaces, int trailingSpaces, int
+                .add(createWhitespaceMinutiae(getWhiteSpaces(leadingSpaces, leadingNewLines)));
+        MinutiaeList newTrailingMinutiaeList = preserveComments(token.trailingMinutiae(), trailingNewLines)
+                .add(createWhitespaceMinutiae(getWhiteSpaces(trailingSpaces, trailingNewLines)));
 
         return token.modify(newLeadingMinutiaeList, newTrailingMinutiaeList);
     }
 
-    private static MinutiaeList modifyMinutiaeList(int spaces, int newLines) {
-        Minutiae minutiae = createWhitespaceMinutiae(getWhiteSpaces(spaces, newLines));
-        return createMinutiaeList(minutiae);
+    private static MinutiaeList preserveComments(MinutiaeList minutiaeList, int trailingNewLines) {
+        MinutiaeList minutiaes = AbstractNodeFactory.createEmptyMinutiaeList();
+        if (minutiaeList.size() > 0) {
+            int count = commentCount(minutiaeList);
+            if (count > 0) {
+                int processedCount = 0;
+                for (int i = 0; i < minutiaeList.size(); i++) {
+                    Minutiae minutiae = minutiaeList.get(i);
+                    minutiaes = minutiaes.add(minutiae);
+                    if (minutiae.kind() == SyntaxKind.COMMENT_MINUTIAE) {
+                        processedCount++;
+                        if (processedCount == count) {
+                            if (trailingNewLines == 0) {
+                                minutiaes = minutiaes.add(AbstractNodeFactory
+                                        .createEndOfLineMinutiae(System.getProperty(LINE_SEPARATOR)));
+                            }
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        return minutiaes;
+    }
+
+    private static int commentCount(MinutiaeList minutiaeList) {
+        int count = 0;
+        for (int i = 0; i < minutiaeList.size(); i++) {
+            if (minutiaeList.get(i).kind() == SyntaxKind.COMMENT_MINUTIAE) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    private static int leadingNewLines(NonTerminalNode parent, Token node) {
+        int count = 0;
+        if (parent != null) {
+            int childLocation = getChildLocation(parent, node);
+            if (parent.children().size() > childLocation + 1) {
+                Token nextToken = getFirstToken(parent.children().get(childLocation + 1));
+                if (nextToken != null && nextToken.containsLeadingMinutiae()) {
+                    MinutiaeList minutiaes = nextToken.leadingMinutiae();
+                    if (commentCount(minutiaes) > 0) {
+                        for (Minutiae minutiae : minutiaes) {
+                            if (minutiae.kind() == SyntaxKind.END_OF_LINE_MINUTIAE) {
+                                count++;
+                            } else {
+                                break;
```

Comment:
Can invert the conditions and do an early exit.

Reply:
Done